rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,535 @@
+
+## Value types
+
+### Author
+Each message will include an `author` key, indicating who performed the given intervention.
+
+- `username` (string): the username of the logged-in user
+- `badgeNumber` (string, optional): the badge number of the logged-in user
+- `location` (Location, optional): the location the logged-in user is managing
+```json
+{
+  "username": "example",
+  "badgeNumber": "123456",
+  "location": {"gtfsId": "place-lake"}
+}
+
+or
+
+{
+  "username": "pswartz",
+  "location": {"gtfsId": "place-matt"}
+}
+```
+
+### Car
+Cars (individual parts of a train consist) will be described by their 4 digit number.
+```json
+{
+  "carNumber": "3800"
+}
+```
+It is represented as an object to provide future extensibility if needed.
+
+### Location
+One of:
+- `{"gtfsId": "<GTFS stop ID>"}`, where the GTFS `location_type` is 0 (stop) or 1 (station)
+- `{"glidesId": "<other>"}`: Other unique identifier for non-revenue locations, internal to Glides
+
+### Operator
+Operators will be described by their badge number:
+```json
+{
+  "badgeNumber": "123"
+}
+```
+It is represented as an object to provide future extensibility if needed.
+
+### Optional T
+A value that can either be an instance of T, `"unmodified"`, or `null`. For example, Optional Car can be
+```json
+{ "carNumber": "3800"}
+# or
+"unmodified"
+# or 
+null
+```
+
+### Time
+A [GTFS Time](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#field-types).
+> Time in the HH:MM:SS format (H:MM:SS is also accepted). The time is measured from "noon minus 12h" of the service day (effectively midnight except for days on which daylight savings time changes occur). For times occurring after midnight, enter the time as a value greater than 24:00:00 in HH:MM:SS local time for the day on which the trip schedule begins.  
+> _Example: `14:30:00` for 2:30PM or `25:35:00` for 1:35AM on the next day._
+
+### Trip Key
+For scheduled trips, Glides does not have the same trip ID used by RTR. For added trips, while an ID is provided, it can be the case that an added trip from the Glides perspective is matched to a GTFS trip, in which case RTR will use the GTFS trip ID in publishing the information.
+
+In order to reduce message duplication, a Trip Key is used to identify both added and scheduled trips.
+
+**Added Trip**
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `glidesId` (string): unique value representing this trip for the given `serviceDate`
+```json
+{"serviceDate": "2023-01-20", "glidesId": "ADDED-123"}
+```
+**Scheduled Trip**
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `scheduledTripId` (string): ID referencing the scheduled trip
+```json
+{"serviceDate": "2023-01-20", "scheduledTripId": "102450"}
```

Comment:
What? Is the idea that every consumer needs to memorize what Glides thinks the schedule is in order to be able to make any sense whatsoever of the data? I don't think that's reasonable. RTR shouldn't need to build a full replica of the Glides database just to be able to understand changes to service.

Reply:
I thought that one of your concerns from last week (about putting operator schedule information in the update messages) was that you didn't want to duplicate the schedule information in the other events. 

I agree that this involves more work on the part of consumers. The alternative (which I prefer) is to duplicate the schedule information in the tripKey, likely in the expanded 5-item format @skyqrose proposed. Let me know which one you prefer.

Reply:
Huh? If I said any such thing, it certainly wasn't in the context of making the trip keys worse. The only thing I remember saying as a concern about putting operator schedule information in the update messages was that it'd be redundant to keep including the scheduled operators every time a trip update was sent. If at some point while I was thinking out loud I said something that gave you the impression that I wanted this, then I apologize. I agree with Sky that service date, start time, start location, end time, end location [should be](https://github.com/mbta/technology-docs/pull/18#discussion_r1089229778) the trip key fields for scheduled trips.